### PR TITLE
salto-1872 - Netsuite: handle order problems in workbooks

### PR DESCRIPTION
**Pull request by SeggevHaimovich**
_Handle meaningless changes in the order of fields inside the xml_

---

_There are changes in the staging env: https://salto-io.slack.com/archives/C02BYQ46T34/p1705107160812559_

---
_Release Notes_: 
_None_

---
_User Notifications_: 
_None_

### CODE REVIEW

File: packages/netsuite-adapter/src/client/deploy_xml_utils.ts [hunk 1]
```diff
@@ -27,6 +27,39 @@ const log = logger(module)
...
+const isDataset = (name:string): boolean =>
+  (name.startsWith('~/Objects/custdataset'))
+
+const isWorkbook = (name:string): boolean =>
+  (name.startsWith('~/Objects/custworkbook'))
+
+const isAnalyticsOrCustCollection = (name: string): boolean =>
+  (isCustCollection(name) || isDataset(name) || isWorkbook(name))
+
+const sortingAnalytics = (a:string, b:string): number => {
```
Comment (dantal4) [inline packages/netsuite-adapter/src/client/deploy_xml_utils.ts:42]: 1. why is this needed? we already order `customTypeInfos` by their dependencies
2. if we need to sort the objects, it's nicer to do it on `customTypeInfos`, that is an array of objects with `typeName` attribute in them, and it's always better than string manipulations.

File: packages/netsuite-adapter/src/filters/analytics_definition_handle.ts [hunk 2]
```diff
@@ -91,9 +91,91 @@ const fetchTransformFunc = async (
...
+    if (Array.isArray(newInstanceValues[arrName])) {
+      adjustWorkbookToolsOrder(newInstanceValues[arrName], oldTools)
+    }
+  })
+}
+
 const createAnalyticsInstance = async (
   instance: InstanceElement,
   analyticsType: ObjectType,
+  elementsSource: ReadOnlyElementsSource | undefined,
```
Comment (dantal4) [inline packages/netsuite-adapter/src/filters/analytics_definition_handle.ts:178]: nit - `elementsSource`is always defined in filters
```suggestion
  elementsSource: ReadOnlyElementsSource,
```

File: packages/netsuite-adapter/src/filters/analytics_definition_handle.ts [hunk 3]
```diff
@@ -91,9 +91,91 @@ const fetchTransformFunc = async (
...
+        })
+      }
+    })
+}
+
+const discardUnrelevantChanges = (
+  newInstanceValues: Values,
+  oldTools: Record<string, Values>,
+): void => {
+  const innerArrayNames = [PIVOTS, CHARTS, DATASET_LINKS]
```
Comment (dantal4) [inline packages/netsuite-adapter/src/filters/analytics_definition_handle.ts:167]: should be defined as a const, probably in `analytics_constants.ts`

File: packages/netsuite-adapter/src/filters/analytics_definition_handle.ts [hunk 4]
```diff
@@ -91,9 +91,91 @@ const fetchTransformFunc = async (
...
+  return _.isEqual(values1, values2)
+}
+
+const adjustWorkbookToolsOrder = (workbookTools: Value[], oldWorkbookToolsRecord: Record<string, Values>): void => {
+  workbookTools
+    .filter(values.isPlainObject)
+    .flatMap(obj => Object.values(obj))
+    .forEach((workbookTool: Values) => {
+      if (_.isArray(workbookTool.datasets)) {
+        workbookTool.datasets.sort()
```
Comment (dantal4) [inline packages/netsuite-adapter/src/filters/analytics_definition_handle.ts:148]: this is a list of strings, right?

File: packages/netsuite-adapter/src/filters/analytics_definition_handle.ts [hunk 5]
```diff
@@ -91,9 +91,91 @@ const fetchTransformFunc = async (
...
+  })
+  return _.isEqual(values1, values2)
+}
+
+const adjustWorkbookToolsOrder = (workbookTools: Value[], oldWorkbookToolsRecord: Record<string, Values>): void => {
+  workbookTools
+    .filter(values.isPlainObject)
+    .flatMap(obj => Object.values(obj))
+    .forEach((workbookTool: Values) => {
+      if (_.isArray(workbookTool.datasets)) {
```
Comment (dantal4) [inline packages/netsuite-adapter/src/filters/analytics_definition_handle.ts:147]: can you use a const for `datasets`?

File: packages/netsuite-adapter/src/filters/analytics_definition_handle.ts [hunk 6]
```diff
@@ -91,9 +91,91 @@ const fetchTransformFunc = async (
...
+        workbookTool.datasets.sort()
+      }
+      if (_.isString(workbookTool[SOAP_SCRIPT_ID]) && workbookTool[SOAP_SCRIPT_ID] in oldWorkbookToolsRecord) {
+        const oldPivot = oldWorkbookToolsRecord[workbookTool[SOAP_SCRIPT_ID]]
+        const innerXmlTitles = [DEFINITION, MAPPING]
+        innerXmlTitles.forEach((xmlTitle: string) => {
```
Comment (dantal4) [inline packages/netsuite-adapter/src/filters/analytics_definition_handle.ts:153]: this logic is very complicated and it's hard to follow... can you implement it using walkOnElement?

File: packages/netsuite-adapter/src/filters/analytics_definition_handle.ts (hunk 4 above)
Comment (SeggevHaimovich) [reply packages/netsuite-adapter/src/filters/analytics_definition_handle.ts:148]: Yes the datasets is a list of strings.
For example:
datasets = [
  "custdataset_dilly_1",
  "custdataset_dilly_2",
]

This is part of what made the difference in the staging

File: packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+const log = logger(module)
+const { awu } = collections.asynciterable
+
+const addWorkbookToolsToRecord = (
+  instance: InstanceElement,
+  workbookToolsRecord: Record<string, Values>,
+  name: string
+): void => {
+  if (Array.isArray(instance.value[name])) {
+    instance.value[name].forEach((tool:unknown) => {
```
Comment (dantal4) [inline packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts:37]:
```suggestion
    instance.value[name].forEach((tool: unknown) => {
```

File: packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+const createRecordOfPivotsChartsAndDsLinks = async (
+  elementsSource: ReadOnlyElementsSource,
+  elemID: ElemID,
+): Promise<Record<string, Values>> => {
+  const oldWorkbookToolsRecord: Record<string, Values> = {}
+  const oldInstance = await elementsSource.get(elemID)
+  if (isInstanceElement(oldInstance)) {
+    addWorkbookToolsToRecord(oldInstance, oldWorkbookToolsRecord, PIVOTS)
+    addWorkbookToolsToRecord(oldInstance, oldWorkbookToolsRecord, CHARTS)
+    addWorkbookToolsToRecord(oldInstance, oldWorkbookToolsRecord, DATASET_LINKS)
```
Comment (dantal4) [inline packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts:60]: nit
```suggestion
    INNER_ARRAY_NAMES.forEach(arrayName => {
      addWorkbookToolsToRecord(oldInstance, oldWorkbookToolsRecord, arrayName)
    })
```

File: packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+
+const adjustWorkbookToolsOrder = (workbookTools: Value[], oldWorkbookToolsRecord: Record<string, Values>): void => {
+  workbookTools
+    .filter(values.isPlainObject)
+    .flatMap(obj => Object.values(obj))
+    .forEach((workbookTool: Values) => {
+      if (_.isArray(workbookTool[DATASETS])) {
+        workbookTool[DATASETS].sort()
+      }
+      if (workbookTool[SOAP_SCRIPT_ID] in oldWorkbookToolsRecord) {
```
Comment (dantal4) [inline packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts:105]: nit - for safety
```suggestion
      if (_.isString(workbookTool[SOAP_SCRIPT_ID]) && workbookTool[SOAP_SCRIPT_ID] in oldWorkbookToolsRecord) {
```

File: packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+        compareAndAssignInnerXmls(workbookTool, oldWorkbookTool)
+      }
+    })
+}
+
+const discardUnrelevantChanges = (
+  newInstanceValues: Values,
+  oldTools: Record<string, Values>,
+): void => {
+  INNER_ARRAY_NAMES.forEach((arrName: string) => {
```
Comment (dantal4) [inline packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts:116]: nit - type announcing is redundant when then type of `INNER_ARRAY_NAMES` is well defined:
```suggestion
  INNER_ARRAY_NAMES.forEach(arrName => {
```

File: packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+      adjustWorkbookToolsOrder(newInstanceValues[arrName], oldTools)
+    }
+  })
+}
+
+const addPermissionToDataset = (
+  dataset1Name: string,
+  dataset2Name: string,
+  datasetMap: Map<string, InstanceElement>,
+): void => {
```
Comment (dantal4) [inline packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts:127]: naming:
1. why is it called "permission"?
2. please use "first" & "second" instead of dataset1/2

File: packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+  if (isSameXmlValues(innerXmlNew, innerXmlOld)) {
+    newTool[innerXmlTitle] = oldTool[innerXmlTitle]
+  }
+}
+
+const adjustWorkbookToolsOrder = (workbookTools: Value[], oldWorkbookToolsRecord: Record<string, Values>): void => {
+  workbookTools
+    .filter(values.isPlainObject)
+    .flatMap(obj => Object.values(obj))
+    .forEach((workbookTool: Values) => {
```
Comment (dantal4) [inline packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts:101]: nit - for safety
```suggestion
    .flatMap(obj => Object.values(obj))
    .filter(values.isPlainObject)
    .forEach((workbookTool: Values) => {
```

File: packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+  oldTools: Record<string, Values>,
+): void => {
+  INNER_ARRAY_NAMES.forEach((arrName: string) => {
+    if (Array.isArray(newInstanceValues[arrName])) {
+      adjustWorkbookToolsOrder(newInstanceValues[arrName], oldTools)
+    }
+  })
+}
+
+const addPermissionToDataset = (
```
Comment (dantal4) [inline packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts:123]: I think that this function requires a comment that explain why are we doing it

File: packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts [hunk 14]
```diff
@@ -0,0 +1,199 @@
...
+  datasetMap: Map<string, InstanceElement>,
+): void => {
+  const firstDataset = datasetMap.get(dataset1Name)
+  const secondDataset = datasetMap.get(dataset2Name)
+  if (
+    isInstanceElement(firstDataset)
+    && isInstanceElement(secondDataset)
+    && Array.isArray(secondDataset.value[DEPENDENCIES]?.[DEPENDENCY])
+  ) {
+    secondDataset.value[DEPENDENCIES]?.[DEPENDENCY]
```
Comment (dantal4) [inline packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts:135]: we shouldn't modify the dependencies list, as it is a part of the element that will be in the XML and will be deployed. we should use a new "made up" field, only so the dependencies mechanism will set a dependency between the datasets, and then delete it before creating the XML.
I think that we should have two functions in `client/utils.ts` for that:
```typescript
const addAdditionalDependency = (instance: InstanceElement, scriptId: string): void => {
  if (instance.value[ADDITIONAL_DEPS] === undefined) {
     instance.value[ADDITIONAL_DEPS] = []
  }
  instance.value[ADDITIONAL_DEPS].push(`[${SCRIPT_ID}=${scriptId}]`)
}

const removeAdditionalDependencies = (instance: InstanceElement): void => {
  delete instance.value[ADDITIONAL_DEPS]
}
```

File: packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts [hunk 15]
```diff
@@ -0,0 +1,199 @@
...
+      ?.push(new ReferenceExpression(firstDataset.elemID, firstDataset.value[SCRIPT_ID], firstDataset))
+  }
+}
+
+const getDatasetLinkList = (workbook: InstanceElement): Values[] =>
+  (Array.isArray(workbook.value[DATASET_LINKS])
+    ? workbook.value[DATASET_LINKS]
+      .filter(values.isPlainObject)
+      .map((obj:Values) => obj[DATASET_LINK])
```
Comment (dantal4) [inline packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts:144]: nit
```suggestion
      .map(obj => obj[DATASET_LINK])
```

File: packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts [hunk 16]
```diff
@@ -0,0 +1,199 @@
...
+      .filter(values.isPlainObject)
+    : [])
+
+const filterCreator: LocalFilterCreator = ({ elementsSource }) => ({
+  name: 'parseAnalytics',
```
Comment (dantal4) [inline packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts:149]: rename?

File: packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+      .filter(change => change.data.after.elemID.typeName === DATASET)
+      .map(getChangeData)
+      .filter(isInstanceElement)
+
+    const datasetMap = new Map<string, InstanceElement>()
+    changedDatasets.forEach(dataset =>
+      datasetMap.set(
+        dataset.value[SCRIPT_ID],
+        dataset
+      ))
```
Comment (dantal4) [inline packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts:173]: nit - 
```suggestion
    const datasetMap = new Map<string, InstanceElement>(
      changedDatasets.map(dataset => [dataset.value[SCRIPT_ID], dataset]
    )
```

File: packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts [hunk 18]
```diff
@@ -0,0 +1,199 @@
...
+
+    const workbooks = changes
+      .filter(isAdditionOrModificationChange)
+      .filter(change => change.data.after.elemID.typeName === WORKBOOK)
+      .map(getChangeData)
+      .filter(isInstanceElement)
+
+    workbooks.forEach(workbook => {
+      getDatasetLinkList(workbook)
+        .forEach((dslink: Values) => {
```
Comment (dantal4) [inline packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts:183]: dsLink is already `Values`
```suggestion
        .forEach(dslink => {
```

File: packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts [hunk 19]
```diff
@@ -0,0 +1,199 @@
...
+          if (Array.isArray(dslink[DATASETS])) {
+            const sortedDatasetArr = [...dslink[DATASETS]].sort()
```
Comment (dantal4) [inline packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts:185]: why do we need to sort it? isn't it sorted in fetch?

File: packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts [hunk 20]
```diff
@@ -0,0 +1,199 @@
...
+            if (dslink[DATASETS].length > 2) {
+              log.debug('There is a datasetLink with more than 2 datasets')
+              for (let i = 0; i < dslink[DATASETS].length - 1; i += 1) {
+                addPermissionToDataset(sortedDatasetArr[i], sortedDatasetArr[i + 1], datasetMap)
+              }
+            }
+            addPermissionToDataset(sortedDatasetArr[0], sortedDatasetArr[1], datasetMap)
```
Comment (dantal4) [inline packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts:192]: I think that this can be DRY, and please use `.reduce(...)` / `.forEach(...)` instead of a for loop.

File: packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts [hunk 21]
```diff
@@ -0,0 +1,199 @@
...
+      .filter(change => change.data.after.elemID.typeName === WORKBOOK)
+      .map(getChangeData)
+      .filter(isInstanceElement)
+
+    workbooks.forEach(workbook => {
+      getDatasetLinkList(workbook)
+        .forEach((dslink: Values) => {
+          if (Array.isArray(dslink[DATASETS])) {
+            const sortedDatasetArr = [...dslink[DATASETS]].sort()
+            if (dslink[DATASETS].length > 2) {
```
Comment (dantal4) [inline packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts:186]: maybe return if `dslink[DATASETS].length === 1` ?

**Review by dantal4 — COMMENTED** [review]
Great work!

File: packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts (hunk 19 above)
Comment (SeggevHaimovich) [reply packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts:185]: not always
we saw that ti keeps happening in staging and if the order was the same in every fetch there was not a repeating problem

File: packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts (hunk 21 above)
Comment (SeggevHaimovich) [reply packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts:186]: The array represents a dataset link, which means there is a link between several datasets...
There shouldn't be any case with 1 dataset in this array.

Comment (dantal4) [reply packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts:186]: but how will the code behave if there's only 1 dataset?

Comment (SeggevHaimovich) [reply packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts:186]: after my change with the reduce there shouldn't be a problem anyway

File: packages/netsuite-adapter/src/filters/analytics_definition_handle.ts [hunk 22]
```diff
@@ -306,6 +308,7 @@ const returnToOriginalShape = async (
   }).parse({ [ROOT]: definitionValues })
 
   return {
+    ..._.omit(instance.value, Object.keys(definitionValues)),
```
Comment (dantal4) [inline packages/netsuite-adapter/src/filters/analytics_definition_handle.ts:311]: in this case, setting explicitly the name, scriptid, dependencies & application_id is redundant, right?

File: packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts [hunk 23]
```diff
@@ -0,0 +1,200 @@
...
+const addDependencyToDataset = (
+  datasetFirstName: string,
+  datasetSecondName: string,
+  datasetMap: Map<string, InstanceElement>,
+): void => {
+  const firstDataset = datasetMap.get(datasetFirstName)
+  const secondDataset = datasetMap.get(datasetSecondName)
+  if (isInstanceElement(firstDataset) && isInstanceElement(secondDataset)) {
+    addAdditionalDependency(secondDataset, firstDataset.value[SCRIPT_ID])
+    // addAdditionalDependencyReference(secondDataset, firstDataset)
```
Comment (dantal4) [inline packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts:136]: leftover?

File: packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts [hunk 24]
```diff
@@ -0,0 +1,200 @@
...
+          if (isStringArray(datasetNamesArray)) {
+            if (datasetNamesArray.length !== 2) {
+              log.debug('There is a datasetLink with different number of datasets than 2: %o', dslink)
+            }
+            datasetNamesArray.reduce((prev, current, index) => {
+              if (index > 0) {
+                addDependencyToDataset(prev, current, datasetMap)
+              }
+              return current
+            }, '')
```
Comment (dantal4) [inline packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts:193]: nit - if you don't pass the initial value (''), then the reduce start running from index 1 (so you won't need the condition):
```typescript
> [1,2,3,4].reduce((x, y, i) => {console.log({x,y,i}); return y})
{ x: 1, y: 2, i: 1 }
{ x: 2, y: 3, i: 2 }
{ x: 3, y: 4, i: 3 }
```

**Review by dantal4 — APPROVED** [review]
💯
